kvserver: avoid rangefeedMu deadlock

stevendanna opened the pull request:
In #132857 we allowed a processor to unregister itself from a scheduler and (more importantly here) to unset itself from the replica when stopping.

Readers should not that the word "register" is used to describe the processor being added to the schedule and the rangefeed registration being added to the processor.

When making this change, we did not remove the other places where we proactively unset the processor from the replica. I am not able to reconstruct why, but my best guess is that when thinking about whether it would be a problem to unset from the replica twice, we were considering whether the function was safe to call twice in terms of removing the wrong processor rather than the locking around rangefeedMu.

But, this change introduces an important point of contention:

1. Before calling p.Register, the thread processing the rangefeed registration locks rangefeedMu on replica r1.

2. A previously enqueued Stop() request for the processor owned by r1 is picked up by the scheduler and executed. It then blocks waiting on rangefeedMu of r1 via a callback that was passed to it at startup.

3. Back in our rangefeed registration thread, to unlock rangefeedMu we need to return from p.Register(). But, p.Register makes a syncronous request whose return requires that either p.stoppedC is closed or the scheduler processes our sync request.

So, how does this _ever_ resolve?

In the happiest case, the processor we are stopping in (2) is the same processor that we are waiting on in (3). In this case, the stop routine closes p.stoppedC before blocking on rangefeedMu. Closing that channel allows p.Register to return and rangefeedMu to eventually be released.

However, some unhappy cases occur if the processor in step 2 (say p1) and step 3 say (p2) are not the same processor. This can occur if p1 was unset from the replica and then a newly created p2 was set on the replica and then the rangefeed registration request arrived. While these are two different processors, they will both interact with the same rangefeedMu from r1.

In this case, p.Register is not unblocked (because it's stoppedC channel won't be closed), so it _requires_ that the scheduler make progress on its request.  But there are at least two cases where it will make no progress:

1. During server shutdown, the worker that handles that shard of processors may have already exited.

2. If the two processors happen to get assigned to the same worker, the worker could be blocked processing the stop request on processor p1 and never make it to p2's register request.

Here, I attempt to resolve this by making the processor's stop function the only place where we remove the processor from the replica.

All other removals should have been in response to a processor that is stopped or stopping and as a result we should still be removing the processor in all of the same conditions, however potentially with a little higher latency. I think the main cause of this latency would be failed rangefeed registration requests that might have otherwise succeed.

The reasoning for why this fixes the bug is that:

1. The only place a blocking call is made while holding rangefeedMu of range r1 is when registering a new rangefeed on its existing processor p1

2. If a concurrent stop request require the rangefeedMu of r1 then it must be the stop request for processor p1 since a stop request is the only thing that can change the processor on r1.

3. A stop request for processor p1 will ensure any currently blocking Register requests for p1 return by closing stoppedC before attempting to unset the replica.

I've re-arranged the code a bit to make (1) more easily reviewable.

Fixes: #144828
Epic: None

Release Note(bug): Fix a bug that could lead to rangefeed for node deadlocks.

stevendanna committed:
kvserver: avoid rangefeedMu deadlock

In #132857 we allowed a processor to unregister itself from a
scheduler and (more importantly here) to unset itself from the
replica when stopping.

Readers should not that the word "register" is used to describe the
processor being added to the schedule and the rangefeed registration
being added to the processor.

When making this change, we did not remove the other places where we
proactively unset the processor from the replica. I am not able to
reconstruct why, but my best guess is that when thinking about whether
it would be a problem to unset from the replica twice, we were
considering whether the function was safe to call twice in terms of
removing the wrong processor rather than the locking around
rangefeedMu.

But, this change introduces an important point of contention:

1. Before calling p.Register, the thread processing the rangefeed
   registration locks rangefeedMu on replica r1.

2. A previously enqueued Stop() request for the processor owned by r1
   is picked up by the scheduler and executed. It then blocks waiting on
   rangefeedMu of r1 via a callback that was passed to it at startup.

3. Back in our rangefeed registration thread, to unlock rangefeedMu we
   need to return from p.Register(). But, p.Register makes a syncronous
   request whose return requires that either p.stoppedC is closed or the
   scheduler processes our sync request.

So, how does this _ever_ resolve?

In the happiest case, the processor we are stopping in (2) is the same
processor that we are waiting on in (3). In this case, the stop
routine closes p.stoppedC before blocking on rangefeedMu. Closing that
channel allows p.Register to return and rangefeedMu to eventually be
released.

However, some unhappy cases occur if the processor in step 2 (say p1)
and step 3 say (p2) are not the same processor. This can occur if p1
was unset from the replica and then a newly created p2 was set on the
replica and then the rangefeed registration request arrived. While
these are two different processors, they will both interact with the
same rangefeedMu from r1.

In this case, p.Register is not unblocked (because it's stoppedC
channel won't be closed), so it _requires_ that the scheduler make
progress on its request.  But there are at least two cases where it
will make no progress:

1. During server shutdown, the worker that handles that shard of
   processors may have already exited.

2. If the two processors happen to get assigned to the same worker,
   the worker could be blocked processing the stop request on processor
   p1 and never make it to p2's register request.

Here, I attempt to resolve this by making the processor's stop
function the only place where we remove the processor from the
replica.

All other removals should have been in response to a processor that is
stopped or stopping and as a result we should still be removing the
processor in all of the same conditions, however potentially with a
little higher latency. I think the main cause of this latency would be
failed rangefeed registration requests that might have otherwise
succeed.

The reasoning for why this fixes the bug is that:

1. The only place a blocking call is made while holding rangefeedMu of
   range r1 is when registering a new rangefeed on its existing processor
   p1

2. If a concurrent stop request require the rangefeedMu of r1 then it
   must be the stop request for processor p1 since a stop request is the
   only thing that can change the processor on r1.

3. A stop request for processor p1 will ensure any currently blocking Register
   requests for p1 return by closing stoppedC before attempting to
   unset the replica.

I've re-arranged the code a bit to make (1) more easily reviewable.

Fixes: #144828
Epic: None

Release Note(bug): Fix a bug that could lead to rangefeed for node
deadlocks.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145956)
<!-- Reviewable:end -->

stevendanna commented on pkg/kv/kvserver/replica_rangefeed.go:
```diff
@@ -590,7 +567,6 @@ func (r *Replica) disconnectRangefeedSpanWithErr(span roachpb.Span, pErr *kvpb.E
 		return
 	}
 	p.DisconnectSpanWithErr(span, pErr)
-	r.maybeDisconnectEmptyRangefeed(p)
```
We may want to replace this with a call to `updateRangefeedFilterLocked()`. But now I'm actually really wondering about the management of this rangefeed filter as well.

stevendanna committed:
kvserver: avoid using stopped rangefeed processor

Previously, we always removed the rangefeed processor from the
replica immediately after issuing a stop request. This was removed to
avoid a deadlock that was introduced.

However, to avoid changing behaviour related to using the processor
for longer than we might have before, we now set a boolean that we can
use to avoid using the replica after it was stopped.

Epic: none

Release note: None
